ENH: Add support for LoRA hotswapping

BenjaminBossan opened the pull request:
# What does this PR do?

LoRA hotswapping has been available in PEFT since 0.15.0. There is already a diffusers
integration (https://github.com/huggingface/diffusers/pull/9453), but the transformers integration was still missing this feature. This PR remedies this. It sticks closely to the diffusers PR, both implementation-wise and API-wise. 

Hotswapping allows to swap different LoRA adapters in-place instead of loading multiple adapters and switching between them. Not only can this be advantageous to safe memory and potentially for quicker loading, the biggest advantage is that if the model is compiled, we can hotswap without triggering recompilation (loading a separate adapter would require recompilation).

There are some caveats to using this feature, most notably that only LoRA is supported. This was fine for diffusers, as it only works with LoRA, but the transformers integration works with other PEFT methods too. However, LoRA should be by far the most common method, so this should be fine for now. This and other caveats have been documented.

Note that testing is not super deep, there could be more edge cases being tested. However, as this is mainly about calling PEFT functionality, which is extensively tested in PEFT, I focused the tests mostly on the integration, with only a couple of tests for the functionality itself (namely the tests that call `_check_model_hotswap`).

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

BenjaminBossan committed:
ENH Add support for LoRA hotswapping

LoRA hotswapping has been available in PEFT since 0.15.0. There is
already a diffusers
integration (https://github.com/huggingface/diffusers/pull/9453), but
the transformers integration was still missing this feature. This PR
remedies this.

Hotswapping allows to swap different LoRA adapters in-place instead of
loading multiple adapters and switchint between them. Not only can this
be advantageous to safe memory and potentially for quicker loading, the
biggest advantage is that if the model is compiled, we can hotswap
without triggering recompilation (loading a separate adapter would
require recompilation).

There are some caveats to using this feature, most notably that only
LoRA is supported. This was fine for diffusers, as it only works with
LoRA, but the transformers integration works with other PEFT methods
too. However, LoRA should be by far the most common method, so this
should be fine for now. This and other caveats have been documented.

BenjaminBossan committed:
Make style

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_41297). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

BenjaminBossan committed:
Merge branch 'main' into enh-support-lora-hotswapping

sayakpaul reviewed (approved):
LGTM 🔥 

Let's maybe update `diffusers` to uplift our current constraint that this feature isn't available for text encoder LoRAs once this is merged.

BenjaminBossan commented on tests/peft_integration/test_peft_integration.py:
```diff
@@ -940,3 +941,175 @@ def test_peft_pipeline_no_warning(self):
+    def tearDown(self):
+        # It is critical that the dynamo cache is reset for each test. Otherwise, if the test re-uses the same model,
+        # there will be recompilation errors, as torch caches the model when run in the same process.
+        torch._dynamo.reset()
```
Done, also added the use of `torch._inductor.utils.fresh_inductor_cache()`.

BenjaminBossan commented on tests/peft_integration/test_peft_integration.py:
```diff
@@ -940,3 +941,175 @@ def test_peft_pipeline_no_warning(self):
+@require_peft
+@require_torch
+@slow
+class PeftHotswapIntegrationTester(unittest.TestCase):
```
Just a note, these are the same decorators as for `PeftIntegrationTester`, but I didn't want to use it here (nor `PeftTesterMixin`), as the hotswap tests don't make use of the test matrix defined there.

BenjaminBossan committed:
Reviewer feedback

sayakpaul commented on docs/source/en/peft.md:
```diff
@@ -151,3 +151,45 @@ model.enable_adapters()
+
+However, there can be situations where recompilation is unavoidable. For example, if the hotswapped adapter targets more layers than the initial adapter, then recompilation is triggered. Try to load the adapter that targets the most layers first. Refer to the PEFT docs on [hotswapping](https://huggingface.co/docs/peft/main/en/package_reference/hotswap#peft.utils.hotswap.hotswap_adapter) for more  ...
+
+> [!Tip]
```
I am referring to the numbers we gathered in https://github.com/huggingface/lora-fast

BenjaminBossan replied:
Done

BenjaminBossan committed:
Reviewer feedback: link benchmark

BenjaminBossan commented:
@ydshieh How to proceed with this PR? Sayak approved but I guess we need a transformers dev's approval too?

BenjaminBossan committed:
Reviewer feedback: rename test class

BenjaminBossan committed:
Merge branch 'main' into enh-support-lora-hotswapping

BenjaminBossan committed:
Reviewer feedback: auto enable hotswap

To make the usage more intuitive, hotswap is now auto-enabled after
calling model.enable_peft_hotswap(). For this, we detect if
enable_peft_hotswap() was called *and* if the adapter being loaded
is *not* the first adapter (because the first adapter cannot be
hotswapped, it needs to be loaded normally).

Cyrilvallez commented on src/transformers/integrations/peft.py:
```diff
@@ -159,12 +161,63 @@ def load_adapter(
                 `find_adapter_config_file` method.
         """
+        from peft import PeftType
+
```
Does that import exist even for the `MIN_VERSION` below? If not, let's import after the check, so that we have gracious crash

Cyrilvallez commented on src/transformers/integrations/peft.py:
```diff
@@ -159,12 +161,63 @@ def load_adapter(
+        if hotswap == "auto":
+            # if user called model.enable_peft_hotswap and this is not the first adapter, enable hotswap
+            hotswap_enabled = getattr(self, "_hotswap_enabled", False)
+            not_first_adapter = bool(self._hf_peft_config_loaded and (adapter_name in self.peft_config))
```
We should not need bool casting here, do we?
```suggestion
            not_first_adapter = self._hf_peft_config_loaded and adapter_name in self.peft_config
```

BenjaminBossan replied:
Right now, `_hf_peft_config_loaded` is a bool, so coercing is not needed. But if its type changes in the future, the type of `not_first_adapter` could also change, so the cast makes this line future proof. Just an example:

```python
>>> _hf_peft_config_loaded = {"foo": 1}
>>> adapter_name = "default"
>>> peft_config = {"default": 2}
>>> not_first_adapter = _hf_peft_config_loaded and adapter_name in peft_config
>>> not_first_adapter
True
>>> 
>>> _hf_peft_config_loaded = {}
>>> not_first_adapter = _hf_peft_config_loaded and adapter_name in peft_config
>>> not_first_adapter
{}
```

As you can see, in the last line, we suddenly have a different type for `not_first_adapter`.

Cyrilvallez commented on src/transformers/integrations/peft.py:
```diff
@@ -261,12 +322,47 @@ def load_adapter(
+
+                prepare_model_for_compiled_hotswap(self, config=peft_config, **self._prepare_peft_hotswap_kwargs)
+                # We only want to call prepare_model_for_compiled_hotswap once
+                self._prepare_peft_hotswap_kwargs = None
```
Should this be under the condition `if not hotswap`? 🤔

Cyrilvallez reviewed (commented):
Very nice! Just mostly a bit concerned about the default which seem to be able to crash currently code that is valid?

BenjaminBossan committed:
Reviewer feedback: move import

BenjaminBossan committed:
Merge branch 'main' into enh-support-lora-hotswapping

BenjaminBossan reviewed (commented):
@Cyrilvallez Thanks a lot for your review. I addressed your comments, please check again.

Cyrilvallez reviewed (approved):
Alright then, thanks for tge added clarifications! And sorry for the delay! Looks like you have a small conflict, but feel free to merge afterwards!

BenjaminBossan committed:
Merge branch 'main' into enh-support-lora-hotswapping
